signer/src/codec.rs

```diff
@@ -73,35 +71,76 @@ pub trait Decode: Sized {
     ///
     /// # Returns
     /// A `Result` which is `Ok` containing the decoded object, or an `Error` if decoding failed.
-    fn decode<R: io::Read>(reader: R) -> Result<Self, Error>;
+    fn decode<R: io::Read>(reader: R) -> Result<Self, CodecError>;
 }
 
-impl<T: serde::Serialize> Encode for &T {
-    fn encode<W: io::Write>(self, writer: W) -> Result<(), Error> {
-        bincode::serialize_into(writer, self)
+impl<T> Encode for T
+where
+    T: ProtoSerializable + Clone,
+    T: Into<<T as ProtoSerializable>::Message>,
+{
+    fn encode_to_vec(self) -> Vec<u8> {
+        let message: <Self as ProtoSerializable>::Message = self.into();
+        prost::Message::encode_to_vec(&message)
     }
 }
 
-impl<T: serde::de::DeserializeOwned> Decode for T {
-    fn decode<R: io::Read>(reader: R) -> Result<Self, Error> {
-        bincode::deserialize_from(reader)
+impl<T> Decode for T
+where
+    T: ProtoSerializable + Clone,
+    T: TryFrom<<T as ProtoSerializable>::Message, Error = Error>,
+{
+    fn decode<R: io::Read>(mut reader: R) -> Result<Self, CodecError> {
+        let mut buf = Vec::new();
+        reader
+            .read_to_end(&mut buf)
+            .map_err(CodecError::DecodeIOError)?;
+
+        let message =
+            <<T as ProtoSerializable>::Message>::decode(&*buf).map_err(CodecError::DecodeError)?;
+
+        T::try_from(message).map_err(|e| CodecError::InternalTypeConversionError(Box::new(e)))
```

Comment:
Oh, I think we should have this trait return the crate error and implement `From` for this the `CodecError`. That we we don't embed the crate error in the `CodecError`. Also, our traits usually return the crate error (for simplicity).